cmd/dra-example-controller/driver.go

```diff
@@ -133,17 +141,85 @@ func (d driver) Allocate(ctx context.Context, claim *resourcev1.ResourceClaim, c
 		err = fmt.Errorf("unknown ResourceClaim.ParametersRef.Kind: %v", claim.Spec.ParametersRef.Kind)
 	}
 	if err != nil {
-		return nil, fmt.Errorf("unable to allocate devices on node '%v': %v", selectedNode, err)
+		return fmt.Errorf("unable to allocate devices on node '%v': %v", selectedNode, err)
 	}
 
 	err = client.Update(&crd.Spec)
 	if err != nil {
-		return nil, fmt.Errorf("error updating NodeAllocationState CRD: %v", err)
+		return fmt.Errorf("error updating NodeAllocationState CRD: %v", err)
 	}
 
 	onSuccess()
 
-	return buildAllocationResult(selectedNode, true), nil
+	return nil
+}
+
+func (d driver) allocateImmediateClaim(claim *resourcev1.ResourceClaim, claimParameters interface{}, class *resourcev1.ResourceClass, classParameters interface{}) (*resourcev1.AllocationResult, error) {
+
+	crdconfig := &nascrd.NodeAllocationStateConfig{
+		Namespace: d.namespace,
+	}
+	crd := nascrd.NewNodeAllocationState(crdconfig)
+
+	client := nasclient.New(crd, d.clientset.NasV1alpha1())
+
+	nasnames, err := client.ListNames()
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving list of NodeAllocationState objects: %v", err)
+	}
+
+	// create claimAllocation
+	ca := &controller.ClaimAllocation{
+		Claim:           claim,
+		ClaimParameters: claimParameters,
+		Class:           class,
+		ClassParameters: classParameters,
+	}
+
+	for _, nodename := range nasnames {
+		if allocationResult, err := d.allocateImmediateClaimOnNode(ca, nodename); err == nil {
+			return allocationResult, nil
+		}
+	}
+
+	return nil, fmt.Errorf("no suitable node found")
+}
+
+func (d driver) allocateImmediateClaimOnNode(ca *controller.ClaimAllocation, nodename string) (*resourcev1.AllocationResult, error) {
```

Comment:
Something doesn't feel right with all the redundancies in this function (especially around creating a new client). It feels like some other abstraction / refactoring would be better, but I don't have a suggestion off the top of my head without digging into the code myself and playing with it.